Improve Security Considerations section (#1625)
moq-wg/moq-transport: draft-ietf-moq-transport.md

  WebTransport: I-D.ietf-webtrans-http3

informative:
  CAT: I-D.ietf-moq-c4m
  PPA: I-D.ietf-moq-privacy-pass-auth

>   I-D.jennings-moq-secure-objects:

Reviewer: Is this right? Didn't we adopt secure objects?

is source authenticity, i.e. to know that the received media objects
are what the original publisher actually published. The information
needing to be source authenticated is not only the encoded media
itself, but also some meta data. For example timestamps are crucial to
understand where on the timeline this media fragment belongs.

The second aspect is content confidentiality, i.e. how to prevent the
relays from having access to the media content itself and minimize what
the relay can deduce about the content. Media object
sizes and traffic patterns do enable some analysis of the
content. Also the track namespace and track name are possible to
analyze and also correlate between end subscribers by relays.

The end-to-end media security is handled by mechanisms external to this
specification. They need to provide source authenticity and
confidentiality. MOQT's media object model does enable both the media
data itself as well as media object headers to be confidentiality and
integrity protected, as well as supporting a class of visible to relay
headers that are integrity protected but not encrypted.

Current proposals for media security include:

>  - An E2EE scheme based on SFRAME: {{I-D.jennings-moq-secure-objects}}.

Reviewer 2: Update reference to draft-ietf-moq-secure-objects-00

# Security Considerations {#security}

MOQT is a protocol used hop-by-hop in a chain between original
publishers to relay, (possibly) relay to relay, and relay to end
subscribers. Thus, the security considerations need to consider first
what happens between two nodes, but also consider the impacts end to
end over several hops of MOQT.

This results in a trust model where on each hop the nodes need to be
securely identified, authorized to use resources of the peer, provide
confidentiality and integrity to prevent third party attacks and limit
monitoring and leakage of privacy sensitive information. The relays
within the chain from original publisher to end subscribers will have

> access to track identifiers as well as the object's content

Reviewer: also metadata

unless it is end-to-end encrypted {{sec-media}}.

> Subscriptions require authorization to prevent unauthorized access to

Reviewer: We use subscriptions symmetrically - a publisher can initiate a subscription also.  Can we say "Publishers serving subscriptions require authorization ..."?

content. Relays MUST verify that subscribers are authorized before

> forwarding content. Subscription requests can carry authorization

Reviewer: ```suggestion
serving objects. Subscription requests can carry authorization
```

tokens to prove the subscriber's right to access specific tracks or
namespaces. Relays that aggregate subscriptions from multiple
downstream subscribers MUST ensure each subscriber is independently

> authorized.

Reviewer: Maybe add something that explains "because there may only be one upstream subscription"

### Subscription Amplification

A malicious subscriber could attempt to overwhelm a publisher or relay
by requesting subscriptions to many tracks simultaneously. Relays
SHOULD implement rate limiting on subscription requests and MAY reject
excessive subscriptions with REQUEST_ERROR. Publishers SHOULD monitor
the number of active subscriptions and enforce limits to prevent
resource exhaustion from a single subscriber or session.

TODO: Describe Cache Poisoning attacks

## Communication Security

MOQT depends on a secure transport to provide confidentiality,
integrity and endpoint authentication between subscriber and
publisher. Implementations use QUIC or WebTransport to fulfill
the basic communication security requirements and these
implementations SHOULD follow best practices for TLS 1.3 and QUIC.
Relays MUST be authenticated to
prevent impersonation of relays.

> It must be noted that the basic security protection offered by QUIC or

Reviewer: Lowercase must

TCP/TLS does not prevent traffic pattern analysis as object
sizes, sizes of request messages etc can make it possible for a third
party observer of the traffic between subscriber and publisher to
identify media content, user patterns and media stream origin.

## Authorization {#sec-authorization}

Authentication of original publishers and end subscribers does not
depend on TLS based mechanisms due to challenges with identifiers and
certificate distribution. Instead they rely primarily on
token based schemes.

> Relays SHOULD be identified using TLS based authentication to prevent

Reviewer: This is redundant with 4433, which is a SHOULD

impersonation attacks. Mutual TLS is expected to be widely used for
node level identification between relays, especially within one

> organization. However, in some deployments mutual TLS may be possible

Reviewer: lowercase may

to use also for end subscribers or original publishers. However, as
only node level authentication is provided, what a particular
identified node is allowed to do is not provided at TLS level.

MOQT has functionality to carry Authorization tokens as message

> parameters when requests are sent. These tokens can be of type the

Reviewer: Subscribe sort of mentioned this above.  Maybe add a section heading and ref from there to here?

service requires to meet its security goals which can be
varied. Therefore, already two variants of authorization tokens have

> been defined for MOQT and more may be defined in the future. The

Reviewer: may

current tokens are Privacy Pass Authentication for Media over QUIC
{{PPA}} and Authentication scheme for MOQT using Common Access Tokens
{{CAT}}.

Tokens are expected to contain information about which actions and

> which resources the end point providing the token is authorized to

Reviewer: endpoint

perform and access. Relays when receiving requests will verify the
token to determine that the request made is authorized.

> ### Replay Attacks

Reviewer: This section feels like maybe it's more related to the token schemes than to MOQT.  Perhaps all we need to say is "this is deferred to specific token schemes, which include requirements for relays processing tokens and requests"

Authorization tokens SHOULD include mechanisms to prevent replay

> attacks. Common Access Tokens {{CAT}} provide replay protection through

Reviewer: Is this SHOULD too strong?  Are we saying everyone uses dpop unless they have a really good reason not to?

DPoP (Demonstrating Proof-of-Possession), which binds tokens to client
cryptographic keys and requires fresh proofs with unique identifiers
(jti) and timestamps for each request. Privacy Pass {{PPA}} tokens
include a client-generated nonce for uniqueness and bind tokens to
their issuance context through a challenge digest. Relays SHOULD track
token identifiers within a configured time window to detect and reject
replayed tokens.

## Media Security  {#sec-media}

> The MOQT hop-by-hop model does provide hop by hop confidentiality and

Reviewer: this has hop-by-hop twice - reads awkwardly

integrity protection of the media objects. However, due to the
hop-by-hop model media objects unless additionally protected are
subject to access by the relay as well as modification intentional as

> well as accidental.

Reviewer: This sentence is very awkwardly constructed - rewrite for clarity

The media objects transported by MOQT in various tracks from various
original publishers are subject to several considerations. The first
is source authenticity, i.e. to know that the received media objects
are what the original publisher actually published. The information
needing to be source authenticated is not only the encoded media
itself, but also some meta data. For example timestamps are crucial to
understand where on the timeline this media fragment belongs.

The second aspect is content confidentiality, i.e. how to prevent the
relays from having access to the media content itself and minimize what
the relay can deduce about the content. Media object
sizes and traffic patterns do enable some analysis of the

> content. Also the track namespace and track name are possible to

Reviewer: This was already stated above

# Security Considerations {#security}

MOQT is a protocol used hop-by-hop in a chain between original
publishers to relay, (possibly) relay to relay, and relay to end
subscribers. Thus, the security considerations need to consider first
what happens between two nodes, but also consider the impacts end to
end over several hops of MOQT.

MoQ uses a trust model where on each hop the nodes need to be
securely identified, authorized to use resources of the peer, provide
confidentiality and integrity to prevent third party attacks and limit
monitoring and leakage of privacy sensitive information. The relays
within the chain from original publisher to end subscribers will have
access to Track names, Track Properties, Object Properties, as well as the object's content
unless it is end-to-end encrypted {{sec-media}}.

Publishers require authorization to prevent unauthorized
subscriptions to content. Relays MUST verify that subscribers

> are authorized before serving objects. Subscription requests can carry

Reviewer: ```suggestion
Publishers, including Relays, require authorization to prevent unauthorized
subscriptions to content. Subscription requests can carry
```

authorization tokens (see {{sec-authorization}}) to prove the
subscriber's right to access specific tracks or namespaces. Relays
that aggregate subscriptions from multiple downstream subscribers MUST
ensure each subscriber is independently authorized, as there may be
only one upstream subscription serving multiple downstream subscribers.

### Subscription Amplification

A malicious subscriber could attempt to overwhelm a publisher or relay
by requesting subscriptions to many tracks simultaneously. Relays
SHOULD implement rate limiting on subscription requests and MAY reject

> excessive subscriptions with REQUEST_ERROR. Publishers SHOULD monitor

Reviewer: ```suggestion
excessive subscriptions with REQUEST_ERROR using the EXCESSIVE_LOAD error code.
Publishers SHOULD monitor
```

# Security Considerations {#security}

MOQT is a protocol used hop-by-hop in a chain between original

> publishers to relay, (possibly) relay to relay, and relay to end

Reviewer 3: ```suggestion
MOQT is a protocol used hop-by-hop between original
publishers to relay, (possibly) relay to relay, and relay to end
```

Reviewer: It is most certainly optional relay-to-relay so we should reflect that?

subscribers. Thus, the security considerations need to consider first
what happens between two nodes, but also consider the impacts end to
end over several hops of MOQT.

This results in a trust model where on each hop the nodes need to be
securely identified, authorized to use resources of the peer, provide
confidentiality and integrity to prevent third party attacks and limit

> monitoring and leakage of privacy sensitive information. The relays

Reviewer 3: ```suggestion
MoQ uses a trust model where on each hop the nodes need to be
securely identified, authorized to use resources of the peer, provide
confidentiality and integrity to prevent third party attacks and limit
monitoring and leakage of privacy sensitive information. The relays
```

within the chain from original publisher to end subscribers will have

> access to track identifiers, metadata, as well as the object's content

Reviewer 3: What's metadata?  Track Properties?

```suggestion
access to Track names, Track Properties, Object Properties, as well as the object's content
```

unless it is end-to-end encrypted {{sec-media}}.

> Publishers serving subscriptions require authorization to prevent

Reviewer 3: We could probably drop this paragraph and merge any non-duplicated content below.

> unauthorized access to content. Relays MUST verify that subscribers

Reviewer 3: ```suggestion
Publishers require authorization to prevent unauthorized
subscriptions to content. Relays MUST verify that subscribers
```

> are authorized before serving objects. Subscription requests can carry

Reviewer 3: Is this Relay specific or true of all Publishers?

Reviewer: maybe merge to prev sentence "Publishers, including Relays, "

authorization tokens (see {{sec-authorization}}) to prove the
subscriber's right to access specific tracks or namespaces. Relays
that aggregate subscriptions from multiple downstream subscribers MUST
ensure each subscriber is independently authorized, as there may be

> only one upstream subscription serving multiple downstream subscribers.

Reviewer 3: ```suggestion
ensure each subscriber is independently authorized.
```

Reviewer: I thought this sentence added some value.

> ### Subscription Amplification

Reviewer 3: Shouldn't we mention forward=0 in this section?

Reviewer: Follow up?

A malicious subscriber could attempt to overwhelm a publisher or relay
by requesting subscriptions to many tracks simultaneously. Relays
SHOULD implement rate limiting on subscription requests and MAY reject

> excessive subscriptions with REQUEST_ERROR. Publishers SHOULD monitor

Reviewer 3: Is there a recommendation on what error code to use?

Reviewer: EXCESSIVE_LOAD?

the number of active subscriptions and enforce limits to prevent
resource exhaustion from a single subscriber or session.

TODO: Describe Cache Poisoning attacks

## Communication Security

MOQT depends on a secure transport to provide confidentiality,
integrity and endpoint authentication between subscriber and
publisher. Implementations use QUIC or WebTransport to fulfill
the basic communication security requirements and these
implementations SHOULD follow best practices for TLS 1.3 and QUIC.
Relays MUST be authenticated to

> prevent impersonation of relays.

Reviewer 3: ```suggestion
Relays MUST use authentication to prevent impersonation.
```

Note that the basic security protection offered by QUIC or TCP/TLS
does not prevent traffic pattern analysis as object sizes, sizes of
request messages etc can make it possible for a third party observer
of the traffic between subscriber and publisher to identify media

> content, user patterns and media stream origin.

Reviewer 3: ```suggestion
does not prevent traffic pattern analysis. Object sizes, sizes of
request messages, etc can make it possible for a third party observer
to identify media content, user patterns and media stream origin.
```

## Authorization {#sec-authorization}

Authentication of original publishers and end subscribers does not

> depend on TLS based mechanisms due to challenges with identifiers and

Reviewer 3: ```suggestion
Authentication of original publishers and end subscribers might not
depend on TLS based mechanisms due to challenges with identifiers and
```

> certificate distribution. Instead they rely primarily on

Reviewer 3: ```suggestion
certificate distribution. They can instead rely on
```

token based schemes.

> Mutual TLS is expected to be widely used for node level identification

Reviewer 3: Do we want a ref for what Mutual TLS is?

Reviewer 3: Also, what is this paragraph saying?  You might want to use Mutual TLS?  Or maybe not?

Reviewer: I think "you might use it, or not, and if you did it doesn't have fine-grained authorization"?

between relays, especially within one organization. However, in some
deployments mutual TLS can also be used for end subscribers or
original publishers. However, as only node level authentication is
provided, what a particular identified node is allowed to do is not
provided at TLS level.

MOQT has functionality to carry Authorization tokens as message
parameters when requests are sent. These tokens can be of type the
service requires to meet its security goals which can be
varied. Therefore, already two variants of authorization tokens have

> been defined for MOQT and more are expected to be defined in the future. The

Reviewer 3: ```suggestion
parameters. These tokens can vary based on the application
requirements. Two variants of authorization tokens have already
been defined for MOQT, and more are expected in the future. The
```

current tokens are Privacy Pass Authentication for Media over QUIC
{{PPA}} and Authentication scheme for MOQT using Common Access Tokens
{{CAT}}.

Tokens are expected to contain information about which actions and
which resources the endpoint providing the token is authorized to
perform and access. Relays when receiving requests will verify the

> token to determine that the request made is authorized.

Reviewer 3: ```suggestion
perform and access. Relays will verify the
token to ensure that the request is authorized.
```

### Replay Attacks

Replay protection for authorization tokens is the responsibility of
the specific token scheme used. Token schemes such as {{CAT}} and
{{PPA}} include requirements for relays when processing tokens and
requests.

## Media Security  {#sec-media}

MOQT provides confidentiality and integrity protection of media

> objects on each hop. However, media objects are accessible to relays,

Reviewer 3: ```suggestion
MOQT uses secure transports that provide confidentiality and integrity
protection. However, media objects are accessible to relays,
```

and are subject to both intentional and accidental modification,
unless they are additionally end-to-end protected.

The media objects transported by MOQT in various tracks from various
original publishers are subject to several considerations. The first
is source authenticity, i.e. to know that the received media objects
are what the original publisher actually published. The information
needing to be source authenticated is not only the encoded media

> itself, but also some meta data. For example timestamps are crucial to

Reviewer 3: ```suggestion
are what the original publisher actually published. In addition to
the media objects, it can also be important to authenticate some
Track and Object Properties. For example, timestamps are crucial to
```

understand where on the timeline this media fragment belongs.

The second aspect is content confidentiality. Beyond direct relay
access to media objects, object sizes and traffic patterns enable
analysis of content. Track namespace and track name can also be
analyzed and correlated between end subscribers by relays.

The end-to-end media security is handled by mechanisms external to this
specification. They need to provide source authenticity and

> confidentiality. MOQT's media object model does enable both the media

Reviewer 3: ```suggestion
confidentiality. MOQT's object model does enable both the object
```

> data itself as well as media object headers to be confidentiality and

Reviewer 3: ```suggestion
data itself as well as Object Properties to be confidentiality and
```

integrity protected, as well as supporting a class of visible to relay

> headers that are integrity protected but not encrypted.

Reviewer 3: ```suggestion
integrity protected. MoQT also supports Object Properties being
integrity protected but not encrypted.
```

Current proposals for media security include:
 - An E2EE scheme based on SFRAME: {{I-D.ietf-moq-secure-objects}}.

> ### Key Management

Reviewer 3: It feels like this paragraph could be replaced with a single sentence in the past section that:

"Secure key distribution for end-to-end encryption is specific to the encryption system and deployment, and outside the scope of this document."
